### PR TITLE
[Nvidia] ASIC Firmware update control

### PR DESCRIPTION
**Pull request by fastiuk**
This PR depends on https://github.com/fastiuk/sonic-swss-common/pull/3

#### Why I did it
* Add an ability to control ASIC FW update
  * Ability to enable/disable FW update
  * Ability to choose FW source: image/user-defined

#### How to verify it
Build SONiC
or
When SONiC was already built:
```
rm -f target/python-wheels/bullseye/sonic_yang_models-1.0-py3-none-any.whl
BLDENV=bullseye make -f Makefile.work target/python-wheels/bullseye/sonic_yang_models-1.0-py3-none-any.whl
rm -f target/sonic-mellanox.bin
make target/sonic-mellanox.bin
```
Login into the switch.
Disable FW auto-update:
```
redis-cli -n 4 HSET "FIRMWARE|asic" "auto_update" "disable"
sudo reboot
```
See the results:
Screenshot 2022-07-03 at 02 20 43
Enable auto-update, change source to user-defined and copy FW image in place:
```
redis-cli -n 4 HSET "FIRMWARE|asic" "auto_update" "enable"
redis-cli -n 4 HSET "FIRMWARE|asic" "default" "user"
sudo mkdir -p /host/mlnx/asic/
sudo cp /tmp/fw-SPC2-rel-29_2010_2270-EVB.mfa /host/mlnx/asic/fw.mfa
sudo reboot
```
Screenshot 2022-07-03 at 03 42 37

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
* Add an ability to control ASIC FW update

#### Link to config_db schema for YANG module changes
[sonic-firmware.yang](https://github.com/fastiuk/sonic-buildimage/tree/firmware-update-control/src/sonic-yang-models/yang-models/sonic-firmware.yang)

#### Submodules PR's :
| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-swss-common | [Firmware update control](https://github.com/fastiuk/sonic-swss-common/pull/3) | GitHub issue/pull request detail |

### CODE REVIEW

**Comment by fastiuk** [comment]
@stepanblyschak please review

File: files/build_templates/init_cfg.json.j2 [hunk 1]
```diff
@@ -109,4 +109,13 @@
...
     }
+{% if sonic_asic_platform == 'mellanox' %}
+    ,
+    "FIRMWARE": {
+        "asic": {
+            "auto_update": "enable",
+            "default": "image"
+        }
+    }
+{% endif %}
```
Comment (stepanblyschak) [inline files/build_templates/init_cfg.json.j2:120]: This design must be reviewed with community to cover all platforms and ASIC vendors. It is not easy to push something to configuration that was designed to work only with specific platforms

File: src/sonic-yang-models/yang-models/sonic-firmware.yang [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    container sonic-firmware {
+
+        container FIRMWARE {
+
+            description "FIRMWARE part of config_db.json";
+
+            container asic {
+
+                leaf auto_update {
+                    type auto-update;
```
Comment (stepanblyschak) [inline src/sonic-yang-models/yang-models/sonic-firmware.yang:40]: Please add description for leafs

File: src/sonic-yang-models/tests/files/sample_config_db.json [hunk 3]
```diff
@@ -1911,6 +1911,13 @@
                 "default_dynamic_th": "0",
                 "over_subscribe_ratio": "0"
             }
+        },
+
```
Comment (stepanblyschak) [inline src/sonic-yang-models/tests/files/sample_config_db.json]: Remove empty line

**Review by stepanblyschak — CHANGES_REQUESTED** [review]
I wonder what is the real world use case for the end user. We know that we **must** have FW version and SDK version aligned. Without a mechanism to upgrade SDK and/or a mechanism to ensure user provided FW is guarateed to work with installed SDK version we create a possibility break the system easily.
If the intent is to allow user have a fix in new FW without waiting for the new SONiC image we can tell the user to replace the FW file under /etc and perform reboot to achive it.
I am also not sure why do we need to control "auto_update". We **must** upgrade FW prior to cold/fast/warm upgrade. If the user configuration disables it we should clearly let user know that these flows will be broken.

**Comment by fastiuk** [comment]
> I wonder what is the real world use case for the end user.

It will allow users to control ASIC firmware and install their own images.

> We know that we **must** have FW version and SDK version aligned. Without a mechanism to upgrade SDK and/or a mechanism to ensure user provided FW is guarateed to work with installed SDK version we create a possibility break the system easily.

1. I am not sure we **must** have versions completely aligned. I think on some systems we can have a little deviation between versions at least minor ones.
2. We don't have that mechanism to guarantee that new FW will work with a current SDK, so it will be user-responsibility to use proper FW.

> If the intent is to allow user have a fix in new FW without waiting for the new SONiC image we can tell the user to replace the FW file under /etc and perform reboot to achive it.

We can but it is not a good user experience. In that case, the user will remove the old FW and won't be able to get it back (iow: it is not straightforward). That feature allows user to choose which FW to use.

> I am also not sure why do we need to control "auto_update". We **must** upgrade FW prior to cold/fast/warm upgrade. If the user configuration disables it we should clearly let user know that these flows will be broken.

That requirement came from another Nvidia OS which allows doing so. 
From user perspective it will allow user to install FW, but not install it right now. Only install when user enable auto-update.

File: files/build_templates/init_cfg.json.j2 (hunk 1 above)
Comment (fastiuk) [reply files/build_templates/init_cfg.json.j2:120]: Of course

File: platform/mellanox/mlnx-fw-upgrade.j2 [hunk 4]
```diff
@@ -250,22 +266,46 @@ function UpgradeFWFromImage() {
...
 ParseArguments "$@"
 
 ExitIfQEMU
+ExitIfAutoUpdateDisabled
 
 WaitForDevice
 
-if [ "${IMAGE_UPGRADE}" != "${YES_PARAM}" ]; then
-    UpgradeFW
+DEFAULT_SOURCE="$(sonic-cfggen -d -v FIRMWARE[\'asic\'][\'default\'])"
```
Comment (stepanblyschak) [inline platform/mellanox/mlnx-fw-upgrade.j2:300]: Also please make sure that we are able to perform reboot when database service is down or config db is not initialized

**Comment by stepanblyschak** [comment]
> > I wonder what is the real world use case for the end user.
> 
> It will allow users to control ASIC firmware and install their own images.
> 
> > We know that we **must** have FW version and SDK version aligned. Without a mechanism to upgrade SDK and/or a mechanism to ensure user provided FW is guarateed to work with installed SDK version we create a possibility break the system easily.
> 
> 1. I am not sure we **must** have versions completely aligned. I think on some systems we can have a little deviation between versions at least minor ones.
> 2. We don't have that mechanism to guarantee that new FW will work with a current SDK, so it will be user-responsibility to use proper FW.
> 
> > If the intent is to allow user have a fix in new FW without waiting for the new SONiC image we can tell the user to replace the FW file under /etc and perform reboot to achive it.
> 
> We can but it is not a good user experience. In that case, the user will remove the old FW and won't be able to get it back (iow: it is not straightforward). That feature allows user to choose which FW to use.
> 
> > I am also not sure why do we need to control "auto_update". We **must** upgrade FW prior to cold/fast/warm upgrade. If the user configuration disables it we should clearly let user know that these flows will be broken.
> 
> That requirement came from another Nvidia OS which allows doing so. From user perspective it will allow user to install FW, but not install it right now. Only install when user enable auto-update.

@fastiuk I suggest to contact @nazariig and understand how this effort relates to [FW util](https://github.com/sonic-net/SONiC/blob/master/doc/fwutil/fwutil.md) and ask to review as he's the code owner for fw upgrade script.

I would also suggest to give a confirmation promt/warning to the user when he does warm/fast reboot when "auto_update" is disabled.

**Pull request closed without merging**